File: RiverBooks.Books/BookEndpoints.cs
using FastEndpoints;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Routing;
using Microsoft.Extensions.DependencyInjection;

namespace RiverBooks.Books;

internal interface IBookService
{
    IEnumerable<BookDto> ListBooks();
}

internal sealed class BookService : IBookService
{
    public IEnumerable<BookDto> ListBooks() =>
    [
        new BookDto(Guid.NewGuid(), "The Fellowship of the Ring", "J.R.R. Tolkien"),
        new BookDto(Guid.NewGuid(), "The Two Towers", "J.R.R. Tolkien"),
        new BookDto(Guid.NewGuid(), "The Return of the Ring", "J.R.R. Tolkien")
    ];
}

public sealed record BookDto(Guid Id, string Title, string Author);

public static class BookEndpoints
{
    public const string GetAll = "/books";

    public static IEndpointRouteBuilder MapBookEndpoints(this IEndpointRouteBuilder app)
    {
        app.MapGet(GetAll, (IBookService bookService) => bookService.ListBooks());

        return app;
    }
}

public sealed class ListBooksResponse
{
    public IEnumerable<BookDto> Books { get; set; }
}

internal class ListBooksEndpoint(IBookService bookService) 
    : EndpointWithoutRequest<ListBooksResponse>
{
    public override void Configure()
    {
        Get("/api/books");
        AllowAnonymous();
    }
    
    public override async Task HandleAsync(CancellationToken token)
    {
        await SendAsync(new ListBooksResponse
        {
            Books = bookService.ListBooks()
        }, cancellation: token);
    }
}

public static class BookServiceCollectionExtensions
{
    public static IServiceCollection AddBookServices(this IServiceCollection services)
    {
        services.AddScoped<IBookService, BookService>();

        return services;
    }
}

File: fsharp/SeaBooks.Books/BookEndpoints.fs
namespace SeaBooks.Books

open System

type BookDto = { Id: Guid; Title: string; Author: string }

module BookModule =

    let listBooks () : BookDto seq =
        seq {
            {
                Id = Guid.NewGuid()
                Title = "The Fellowship of the Ring"
                Author = "J.R.R. Tolkien"
            }

            {
                Id = Guid.NewGuid()
                Title = "The Two Towers"
                Author = "J.R.R. Tolkien"
            }

            {
                Id = Guid.NewGuid()
                Title = "The Return of the Ring"
                Author = "J.R.R. Tolkien"
            }
        }

    type BookService() =
        member _.ListBooks() = listBooks ()

module BookEndpoints =

    open System.Runtime.CompilerServices
    open Microsoft.Extensions.DependencyInjection
    open Microsoft.AspNetCore.Http
    open Giraffe
    open Giraffe.EndpointRouting
    open BookModule

    [<Extension>]
    type IServiceCollectionExtensions =
        [<Extension>]
        static member inline AddBookService(services: IServiceCollection) : IServiceCollection =
            services.AddScoped<BookService>(fun sp -> BookService())

    let listBooksHandler =
        fun (next: HttpFunc) (ctx: HttpContext) ->
            let bookService = ctx.GetService<BookService>()
            bookService.ListBooks() |> ctx.WriteJsonAsync

    let bookEndpoints = [ GET [ route "/books" listBooksHandler ] ]
